Android: add better nullability checks for nullability annotations added in NDK 26

finagolfin opened the pull request:
This is needed because [Bionic recently added a bunch of these annotations](https://android.googlesource.com/platform/bionic/+/25725717861813c18189c2699cf663bdb984d1b9%5E%21/#F0). I made sure this pull doesn't break anything by testing most of it with the previous NDK 25c also. I used this patch with others to build the Swift toolchain for my Android CI, finagolfin/swift-android-sdk#122, and [the Termux app for Android](https://github.com/termux/termux-packages/commit/d10fd452662d3c39c0f19a901433bce8711b4688#diff-b8f3b540c354923485cb06eba2d66237536f89e88fd51e170495ddd8f5d32622R44), which now uses NDK 26b.

@drodriguez or @compnerd, please review.

finagolfin commented on Sources/Foundation/FileHandle.swift:
```diff
@@ -310,9 +310,15 @@ open class FileHandle : NSObject {
+                        munmap(buffer, length)
+                    }
+                  #else
                     return NSData.NSDataReadResult(bytes: data!, length: mapSize) { buffer, length in
```
As shown in my Bionic link above, `mmap()` now returns `_Nonnull`, so this force-unwrap fails.

parkera replied:
It feels rather unfortunate to ifdef around this here instead of at the place we call `mmap`.

finagolfin commented on Sources/Foundation/FileHandle.swift:
```diff
@@ -307,10 +307,15 @@ open class FileHandle : NSObject {
                 // Filesizes are often 64bit even on 32bit systems
                 let mapSize = min(length, Int(clamping: statbuf.st_size))
+              #if os(Android)
+                // Bionic mmap() now returns _Nonnull, so the force unwrap below isn't needed.
```
@parkera, is this the change you had in mind?

finagolfin commented on Sources/Foundation/FileManager+POSIX.swift:
```diff
@@ -1085,10 +1091,18 @@ extension FileManager {
             do {
                 guard fm.fileExists(atPath: _url.path) else { throw _NSErrorWithErrno(ENOENT, reading: true, url: url) }
                 _stream = try FileManager.default._fileSystemRepresentation(withPath: _url.path) { fsRep in
+                    #if os(Android)
```
Maybe there is, I just don't know how to do it. This is simply the best I came up with after seeing these Swift type-casting issues for the first time, and I wanted to ask if there was a better way.

@glbrntt, you recently added similar code to NIO, wdyt?

glbrntt replied:
NIO wraps all of the syscalls and does any platform-specific shenanigans within the wrapper.

finagolfin replied:
OK, but that will still require adapting these types for nullability annotations: see [my recent patch of your new NIO code for Android](https://github.com/finagolfin/swift-android-sdk/blob/main/swift-nio-filesystem-ndk26.patch#L56).

glbrntt replied:
re: your patch, does using an implicitly unwrapped optional work for both cases? e.g. `[UnsafeMutablePointer<CInterop.PlatformChar>!]`

finagolfin replied:
I get the following error with Swift 5.9.2 on Android AArch64 if I add that to the type signature of `libc_fts_open()`:
```
swift-nio/Sources/NIOFileSystem/Internal/System Calls/Syscalls.swift:407:14: error: using '!' is not allowed here; perhaps '?' was intended?
    _ path: [UnsafeMutablePointer<CInterop.PlatformChar>!],
             ^                                          ~
                                                        ?
error: fatalError
```

finagolfin replied:
@glbrntt, is that type what you had in mind, or simply unwrapping the arguments when they are passed in inside `ftsOpen()`?

compnerd replied:
I think that I approached it similarly to what was being suggested - rebinding the memory around the call site.

finagolfin commented:
@parkera, any further feedback here?

compnerd commented on Sources/Foundation/FileManager+POSIX.swift:
```diff
@@ -741,32 +741,38 @@ extension FileManager {
                 return
             } else if errno == ENOTEMPTY {
+                #if os(Android)
+                let ps = UnsafeMutablePointer<UnsafeMutablePointer<Int8>>.allocate(capacity: 2)
```
Do we really need to duplicate this or can we get away with a rebinding of the memory?

finagolfin replied:
No longer needed after your swiftlang/swift#74829, so removed in #5010.

compnerd commented on Sources/Foundation/FileManager+POSIX.swift:
```diff
@@ -741,32 +741,38 @@ extension FileManager {
                 ps.deallocate()
 
-                if stream != nil {
+                if let openStream = stream {
```
Why not simply `if let stream`?

finagolfin replied:
Good point, removed in #5010, you can add it in a version of #4889 for 6.0.

compnerd commented on Sources/Foundation/FileManager+POSIX.swift:
```diff
@@ -1315,7 +1329,7 @@ extension FileManager {
                 return newItemURL.withUnsafeFileSystemRepresentation { (newItemFS) -> Int32? in
                     // This is an atomic operation in many OSes, but is not guaranteed to be atomic by the standard.
-                    if rename(newItemFS, originalFS) == 0 {
+                    if let newFS = newItemFS, let origFS = originalFS, rename(newFS, origFS) == 0 {
```
Why not simply force unwrap?

finagolfin replied:
That's better, dropped this in #5010, you can do that in a version of #4889 for 6.0.

compnerd commented on Sources/Foundation/Host.swift:
```diff
@@ -25,7 +25,8 @@ import WinSDK
     // getnameinfo uses size_t for its 4th and 6th arguments.
     private func getnameinfo(_ addr: UnsafePointer<sockaddr>?, _ addrlen: socklen_t, _ host: UnsafeMutablePointer<Int8>?, _ hostlen: socklen_t, _ serv: UnsafeMutablePointer<Int8>?, _ servlen: socklen_t, _ flags: Int32) -> Int32 {
-        return Glibc.getnameinfo(addr, addrlen, host, Int(hostlen), serv, Int(servlen), flags)
+        guard let saddr = addr else { return -1 }
```
I think that I prefer my approach of changing the signature of the private func.

finagolfin replied:
Sorry about the delay, been meaning to get to this, hopefully this week.

finagolfin replied:
Changing the signature would work too, dropped this in #5010, you can submit it in a version of #4889 for 6.0.

finagolfin commented:
Now that `swift-foundation` was merged, this pull no longer applies cleanly to that rewrite. Instead, I have modified or removed portions of this pull based on the review comments here and applied it to the `release/6.0` branch in #5010 instead, where `swift-foundation` will presumably not be merged also.

Once that's in 6.0, I will rework this pull for `swift-foundation` and make the necessary changes here also. My answers to the review comments above will now address what I changed for #5010.

finagolfin commented:
Some of these are no longer needed after the re-core and other changes, others have been submitted again in #5024 and my pull for that, hyp#1.
